src/plugins/data_source/server/client/configure_client.ts

```diff
@@ -117,52 +175,75 @@ export const getCredential = async (
  * @returns child client.
  */
 const getQueryClient = async (
-  rootClient: Client,
-  dataSource: DataSourceAttributes,
+  dataSourceAttr: DataSourceAttributes,
+  addClientToPool: (endpoint: string, authType: AuthType, client: Client | LegacyClient) => void,
   cryptography?: CryptographyServiceSetup,
+  rootClient?: Client,
   requireDecryption: boolean = true
 ): Promise<Client> => {
-  const authType = dataSource.auth.type;
+  const {
+    auth: { type },
+    endpoint,
+    lastUpdatedTime,
+  } = dataSourceAttr;
 
-  switch (authType) {
+  switch (type) {
     case AuthType.NoAuth:
-      return rootClient.child();
+      addClientToPool(endpoint, type, rootClient!);
+      return rootClient!.child();
 
     case AuthType.UsernamePasswordType:
+      addClientToPool(endpoint, type, rootClient!);
       const credential = requireDecryption
-        ? await getCredential(dataSource, cryptography!)
-        : (dataSource.auth.credentials as UsernamePasswordTypedContent);
-      return getBasicAuthClient(rootClient, credential);
+        ? await getCredential(dataSourceAttr, cryptography!)
+        : (dataSourceAttr.auth.credentials as UsernamePasswordTypedContent);
+      return getBasicAuthClient(rootClient!, credential);
+
+    case AuthType.SigV4:
+      const awsCredential = requireDecryption
+        ? await getAWSCredential(dataSourceAttr, cryptography!)
+        : (dataSourceAttr.auth.credentials as SigV4Content);
+
+      const awsClient = rootClient ? rootClient : getAWSClient(awsCredential, endpoint);
+      addClientToPool(endpoint + lastUpdatedTime, type, awsClient);
```

Comment:
maybe we should add data source id in the key as well? Or can we add the access key into cache so that we can verify if the credentials has been updated? Only use lastUpdateTime seems not reasonable

Reply:
make sense, update key to be `url + dataSourceId + lastUpdatedTime`